File: 2025-03-March-GFG-POTD/March 06 - Longest Common Subsequence/Solution.py
class Solution:
    def lcs(self, s1: str, s2: str) -> int:
        m, n = len(s1), len(s2)
        dp = [[0] * (n + 1) for _ in range(m + 1)]

        # Fill the DP table using bottom-up approach
        for i in range(1, m + 1):
            for j in range(1, n + 1):
                if s1[i - 1] == s2[j - 1]:
                    dp[i][j] = 1 + dp[i - 1][j - 1]
                else:
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1])

        return dp[m][n]

# Driver Code
if __name__ == "__main__":
    test = int(input())
    for _ in range(test):
        s1 = input().strip()
        s2 = input().strip()
        obj = Solution()
        print(obj.lcs(s1, s2))
        print("~")

File: 2025-03-March-GFG-POTD/March 06 - Longest Common Subsequence/Solution.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    int lcs(string s1, string s2) {
        int m = s1.length();
        int n = s2.length();
        vector<vector<int>> dp(m + 1, vector<int>(n + 1, 0));

        // Fill the DP table using bottom-up approach
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (s1[i - 1] == s2[j - 1]) {
                    dp[i][j] = 1 + dp[i - 1][j - 1];
                } else {
                    dp[i][j] = max(dp[i - 1][j], dp[i][j - 1]);
                }
            }
        }
        return dp[m][n];
    }
};

int main() {
    int test;
    cin >> test;
    while (test--) {
        string s1, s2;
        cin >> s1 >> s2;
        Solution obj;
        cout << obj.lcs(s1, s2) << endl;
        cout << "~" << endl;
    }
    return 0;
}
